### ISSUE TITLE
Lists as first-class citizens

### ISSUE THREAD

**Issue by mielvds** [w3c/N3#12]
Let me steal an example from @doerthe here:
```
:bob :favouriteFood (:ice-cream :spaghetti :hamburgers).
```
is different from
```
    :bob :favouriteFood _:b1 rdf:first :ice-cream .
    _:b1 rdf:rest _:b2 .
    
_:b2 rdf:first :spaghetti .
    _:b2 rdf:rest _:b3 .

    _:b3 rdf:first :hamburgers .
    _:b3 rdf:rest rdf:nil .
```

Turtle specifies the `(...) ` syntax as an [RDF collection](https://www.w3.org/TR/turtle/#collections), where they are definitely the same.
Thus:
1) Where are they conflicting? What are the differences?
2) Does this cause conflicts with Turtle/RDF? possibly for resolving #1 
3) How can we add compatibility? What are the semantics? Can we keep the same syntax?

**Comment by william-vw** [comment]
It seems that the [spec ](https://www.w3.org/TeamSubmission/n3/#lists) is leaving it up to implementations on how to treat lists (?)

> Implementations may treat list as a data type rather than just a ladder of rdf:first and rdf:rest properties. The use of rdf:first and rdf:rest can be regarded as a reification of the list datatype.

The latter part seems to imply that an N3 list can be readily converted into RDF by using a first/rest ladder.

**Comment by doerthe** [comment]
@mielvds 
First from a practical perspective: I first realised that there was a difference between RDF lists and N3 lists when someone sent me an owl ontology generated using protege for further reasoning. All lists in the ontology were stored as first-rest-pairs and when I wanted to apply rule for OWL-RL reasoning it did not immediately work. It was no problem to write N3 rules to translate the first-rest-ladders to N3 lists.

 So, you will get conflicts if you have the first-rest-lists in an rdf-document  which you want to use in N3 but then can be fixed by adding rules. Of course, N3 lists should also be valid in RDF since N3 uses one of the two possible ways of RDF to use lists.

For theoretical purposes it is also much easier if we have lists as "first-class-citizens" instead of the nested notation because we are not "misusing" blank nodes here. So, there is no need to explain why declaring that there exists a list is the same as having the list in this context (the problem is similar to reification).

@william-vw  Yes, it can be converted. But this is already a consequence of the definitions done in RDF. If the two representations from above need to be the same in RDF, then it should not be a problem to translate one into the other. 
If you want to see a concrete translation I can provide some rules here.

**Comment by william-vw** [comment]
@doerthe I imagine that whether or not first-rest collection ladders work in N3 would depend on the N3 parser. IMO it should be transparent whether first-rest ladders or list datatypes are being used. (I can't imagine any use case where a developer would want to work with the _monstrosities_ that are first-rest ladders instead of "proper" lists as they are found, well, anywhere else.) For sure, the semantics of the conversion could be formally defined by a set of translation rules in both directions (and usable in lieu of parser support).

Not sure what you mean by one of two possible ways:

> Of course, N3 lists should also be valid in RDF since N3 uses one of the two possible ways of RDF to use lists.

**Comment by doerthe** [comment]
> Not sure what you mean by one of two possible ways:
> 
> > Of course, N3 lists should also be valid in RDF since N3 uses one of the two possible ways of RDF to use lists.

Ah, sorry. Here I simply meant that in RDF 

Option 1:
`:bob :favouriteFood (:ice-cream :spaghetti :hamburgers).`

and

Option 2:
```
    :bob :favouriteFood _:b1 rdf:first :ice-cream .
    _:b1 rdf:rest _:b2 .
    
_:b2 rdf:first :spaghetti .
    _:b2 rdf:rest _:b3 .

    _:b3 rdf:first :hamburgers .
    _:b3 rdf:rest rdf:nil .
```
should be equivalent and I just called option 1and option 2  together the "two possible ways" to represent a list. Sorry for the confusion, I could have been clearer.

**Comment by mielvds** [comment]
> difference

**Comment by koo5** [comment]
i have to kind-of disagree that there is "a difference between RDF lists and N3 lists". In our reasoner, N3 lists are translated into RDF ladders on input, and these ladders (hopefully) obey all expected semantics thanks to AE rules. 

But to make this work as expected, while following the limit of one existential in rule head for AE rules, we expand N3 lists in rule heads into ladders in rule bodies, although that seems to be a cosmetic change in semantics. We could avoid this nonstandard behavior if N3 had a way to refer to the subject of the first triple of the ladder form, ie: {?List1 :x :y} <= {(:a :b :c} :z :z}, where List1 would refer to (:a :b :c). (note that ? doesn't signify an universal in our system)

Each reasoner brings it's own semantics to complement N3 or RDF semantics, so i would totally appreciate a specification of what we expect from lists, but this distinction between N3 lists and RDF ladders seems maybe superficial. I will also argue that it might make little sense to post-hoc specify semantics of N3 lists based on existing implementations, because each implementation probably does something a little different. 

So maybe we can specify a way that N3 list syntax actually should be expanded into the ladder form (trivial), and separately of that, specify what we expect implementations to do with these ladders.

**Comment by doerthe** [comment]
> i have to kind-of disagree that there is "a difference between RDF lists and N3 lists".

OK, my statement was based on what is mentioned above: the specifications ([team submission](https://www.w3.org/TeamSubmission/n3/#lists) and [N3 paper](https://www.csee.umbc.edu/courses/graduate/691/spring12/03/papers/n3logic-tplp.pdf)) say that implementers are allowed to understand lists as "first-class-citizens". 

> In our reasoner, N3 lists are translated into RDF ladders on input, and these ladders (hopefully) obey all expected semantics thanks to AE rules.
> 
Sorry for my question, but: what does AE stand for?

Do you do that internally or does the user also see that and can use that notation? 


> But to make this work as expected, while following the limit of one existential in rule head for AE rules, we expand N3 lists in rule heads into ladders in rule bodies, although that seems to be a cosmetic change in semantics. We could avoid this nonstandard behavior if N3 had a way to refer to the subject of the first triple of the ladder form, ie: {?List1 :x :y} <= {(:a :b :c} :z :z}, where List1 would refer to (:a :b :c). (note that ? doesn't signify an universal in our system)
> 

Can you explain that in more detail?
Do you simply want to have names for each sub-list? 

So, with URIs:

```
    :bob :favouriteFood  :uniquiURI1.
              :uniquiURI1   rdf:first :ice-cream .
              :uniquiURI1  rdf:rest  :uniquiURI2 .

:uniquiURI2 rdf:first :spaghetti .
    :uniquiURI2 rdf:rest :uniquiURI2.

   :uniquiURI3 rdf:first :hamburgers .
    :uniquiURI3 rdf:rest rdf:nil .
```
That since that is a bad solution you replaced these unique URIs by variables which have no scoping? Or what did you do? 

To me, it is not clear why we need that in the first place, why can't we just say:

 ```
    :bob :favouriteFood (:ice-cream :spaghetti :hamburgers). ```

`(:ice-cream :spaghetti :hamburgers)` is already kind of a unique name and we even know that all implementations use the same unique name. If we go to model theory, this list itself is then an element of the domain of discourse.

If you still want to state the other triples you could do so by saying.

``` 
             (:ice-cream :spaghetti :hamburgers)   rdf:first :ice-cream .
             (:ice-cream :spaghetti :hamburgers)  rdf:rest  (:spaghetti :hamburgers) .

             (:spaghetti :hamburgers) rdf:first :spaghetti .
             (:spaghetti :hamburgers) rdf:rest ( :hamburgers).

             (:hamburgers) rdf:first :hamburgers .
             (:hamburgers) rdf:rest rdf:nil .
```

Can you also  give me the link to your reasoner so that I can "play" with it to better understand.

> Each reasoner brings it's own semantics to complement N3 or RDF semantics, so i would totally appreciate a specification of what we expect from lists, but this distinction between N3 lists and RDF ladders seems maybe superficial. I will also argue that it might make little sense to post-hoc specify semantics of N3 lists based on existing implementations, because each implementation probably does something a little different.
> 
Yes, you are right, we should not base the semantics we specify on the existing implementations (of course whatever we decide should be implementable and there you hopefully will stop us whenever this is not the case). I just based what I did so far on the original specification of N3 and how I understand it. Whether or not we should follow that is of course another discussion. I just see more problems in making lists rdf:first-rdf:rest-ladders than I see benefits. 

> So maybe we can specify a way that N3 list syntax actually should be expanded into the ladder form (trivial), and separately of that, specify what we expect implementations to do with these ladders.

To agree or disagree I would like to understand the problem a little bit better. So, can you describe your example above in more detail (I will ask a lot of questions, because I really want to know :) ) and then we can have a look how we can specify that. Would you like to see some kind of function as it is done in prolog?

**Comment by gkellogg** [comment]
> On Jun 25, 2019, at 7:26 AM, doerthe <notifications@github.com> wrote:
> 
> i have to kind-of disagree that there is "a difference between RDF lists and N3 lists".
> 
> OK, my statement was based on what is mentioned above: the specifications (team submission <https://www.w3.org/TeamSubmission/n3/#lists> and N3 paper <https://www.csee.umbc.edu/courses/graduate/691/spring12/03/papers/n3logic-tplp.pdf>) say that implementers are allowed to understand lists as "first-class-citizens".
> 
> In our reasoner, N3 lists are translated into RDF ladders on input, and these ladders (hopefully) obey all expected semantics thanks to AE rules.
> 
> Sorry for my question, but: what does AE stand for?
> 
> Do you do that internally or does the user also see that and can use that notation?
> 
My implementation also uses RDF ladders, and can accept it in the notation as well. So far, with no reasoning issues.
> But to make this work as expected, while following the limit of one existential in rule head for AE rules, we expand N3 lists in rule heads into ladders in rule bodies, although that seems to be a cosmetic change in semantics. We could avoid this nonstandard behavior if N3 had a way to refer to the subject of the first triple of the ladder form, ie: {?List1 :x :y} <= {(:a :b :c} :z :z}, where List1 would refer to (:a :b :c). (note that ? doesn't signify an universal in our system)
> 
> Can you explain that in more detail?
> Do you simply want to have names for each sub-list?
> 
> So, with URIs:
> 
>     :bob :favouriteFood  :uniquiURI1.
>               :uniquiURI1   rdf:first :ice-cream .
>               :uniquiURI1  rdf:rest  :uniquiURI2 .
> 
> :uniquiURI2 rdf:first :spaghetti .
>     :uniquiURI2 rdf:rest :uniquiURI2.
> 
>    :uniquiURI3 rdf:first :hamburgers .
>     :uniquiURI3 rdf:rest rdf:nil .
> That since that is a bad solution you replaced these unique URIs by variables which have no scoping? Or what did you do?
> 
The subject of the ladder triples need to be blank nodes, not IRIs.
> To me, it is not clear why we need that in the first place, why can't we just say:
> 
>    :bob :favouriteFood (:ice-cream :spaghetti :hamburgers). ```
> 
> `(:ice-cream :spaghetti :hamburgers)` is already kind of a unique name and we even know that all implementations use the same unique name. If we go to model theory, this list itself is then an element of the domain of discourse.
> 
> If you still want to state the other triples you could do so by saying.
> 
>          (:ice-cream :spaghetti :hamburgers)   rdf:first :ice-cream .
>          (:ice-cream :spaghetti :hamburgers)  rdf:rest  (:spaghetti :hamburgers) .
> 
>          (:spaghetti :hamburgers) rdf:first :spaghetti .
>          (:spaghetti :hamburgers) rdf:rest ( :hamburgers).
> 
>          (:hamburgers) rdf:first :hamburgers .
>          (:hamburgers) rdf:rest rdf:nil .

That doesn’t quite work. The first statement with rdf:first :ice-cream just create a duplicated triple. The second statement with rdf:rest ends up creating a separate list. You’d need to so something more like the following:

	:bob :favoriteFood _:lh .
	_:lh rdf:first :ice-cream .
	_:lh rdf:rest (:spaghetti :hambergers) .

Gregg

**Comment by mielvds** [comment]
> That doesn’t quite work. The first statement with rdf:first :ice-cream just create a duplicated triple.

I don't see what that would be if (:ice-cream :spaghetti :hamburgers)  serves as an identifier. I see no difference with the blanknode example in that regard. And that also depends on the semantics you give it right?

**Comment by gkellogg** [comment]
> > That doesn’t quite work. The first statement with rdf:first :ice-cream just create a duplicated triple.
> 
> I don't see what that would be if (:ice-cream :spaghetti :hamburgers) serves as an identifier. I see no difference with the blanknode example in that regard. And that also depends on the semantics you give it right?

As RDF, `(:ice-cream :spaghetti :hamburgers)` when used as a subject, is the same as using the blank node at the head of the ladder as the subject. Clearly, there are subtle differences here with lists as first-class citizens, but but reasoning can work.

It’s similar if you use a formula as a subject or object: `{ [a rdf:Graph] } a log:truth` could be treated equivalently, if a formula in subject or object position is equivalent to a blank node graph name associated with that formula.

**Comment by ericprud** [comment]
I have PR's on N3.js to [sequester lists](https://github.com/rdfjs/N3.js/pull/179) and [enable N3Writer to serialize sequestered lists](https://github.com/rdfjs/N3.js/pull/180). The strategy was to make sure that any list would either be a first-class object or a first/rest ladder, but never both. (I guess a graph with both would have a [lean representation](https://www.w3.org/TR/rdf11-mt/#dfn-lean) which would remove duplicates but I don't see a lot of leaning in RDF processing.)

**Comment by william-vw** [comment]
Resurrecting this issue for a minute - I think we've actually discussed this on the semantic web mailing list; as per the [team submission](https://www.w3.org/TeamSubmission/n3/#lists), some axioms would need to be added to make N3 collections fully equivalent to RDF first/rest ladders.

If we accept this equivalency (with the above caveat), then the N3 spec could highlight this equivalence (as was done in the team submission), and recommend that implementations support them as first-class citizens - but they should support both RDF first-rest ladders and N3 collections. 

Thoughts?

**Comment by william-vw** [comment]
At some point (meeting notes: 19th June, 2019) we seemed to agree to keep N3 lists separate from first/rest ladders. So it seems that we haven't really reached a consensus on this issue. We should likely discuss this in a next meeting.

**Comment by koo5** [comment]
@doerthe 
> when I wanted to apply rule for OWL-RL reasoning it did not immediately work

I'm still confused by this. Could you point me to the exact reasoner/testcase or something that would show how "n3 lists" and rdf ladders were treated differently please. I mean, from my experience, the duality totally invites a different treatment by reasoner, but i've done my homework to demonstrate to myself that such duality is incorrect. So let's discuss case-by-case

**Comment by gkellogg** [comment]
I worked for some time trying to get first/rest ladders to work, but basically ran into the following problem: when used as an argument to builtin, the list does not necessarily quantify a list in the knowledge base, but when devolved into first/rest statements, those statements exist outside of a builtin argument, and do quantify list elements in the knowledge base, so it leads to matching partial lists that exist in the knowledge base (outermost graph), but when treated as a first-class resource, it doesn't.

Keeping them as first-class resources prevents unwanted quantification, and for me, anyway, really improves query performance.

Note that a list used for non-builtin triples within a formula would, in my interpretation, existentially match a corresponding list in the knowledge base.

(I believe formulae end up working the same way).

**Comment by pchampin** [comment]
@gkellogg can you give an example of the problem you are describing? I find it hard to follow...

**Comment by gkellogg** [comment]
I'd need to reverse some of my changes to find something that fails, but consider the following:

```turtle
{ (1 2 3) list:last ?x } => { :lastMember :is ?x } .
```

With the first-class model, it's clear that `?x` would bind to `3`. But, if stated as follows:

```turtle
{
  _:l3: rdf:first 3; rdf:rest rdf:nil .
  _:l2 rdf:first 2; rdf:rest _:l3 .
  _:l1 rdf:first 1; rdf:rest _:l2; list:last ?x
} => { :lastMember :is ?x } .
```

This would only bind `?x` to `3`, if the knowledge base included a list, which at least ended with `(2 3)`, as `_:l2` and `_:l3` are quantified, and a solution is only found if there is something in the knowledge base that can be bound. But, using first-class lists, there is no existential quantifier involved, so the knowledge base can be empty. (However, if `(1 2 3)` existed as a subject or object of a non-built-in statement, it would be quantified to match an equivalent list in the knowledge base).

Of course, that's just my interpretation, based on my possibly flawed understanding of the logic, but consistent with my own implementation. In my case, I query the knowledge base using the triples contained in the antecedent formula that do not use built-ins (considering `rdf:first` and `rdf:rest` not to be built-ins). Finding no solution, the `list:last` would not be invoked.

**Comment by josd** [comment]
Well, just like list:last I implemented rdf:first also a built-in, but one
that you can extend with your own triples
Remember last Monday how we argued against such extensions like for
instance (1 1) math:sum 3 but this is no
problem for the built-ins rdf:first, rdf:rest, log:implies, etc, ...
Well, to me it smells like a use versus mention bug, but one that we can
deal with esp. because they are built-ins.

Now could you try the following test case:

GIVEN

> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> @prefix list: <http://www.w3.org/2000/10/swap/list#>.
> @prefix : <http://example.org/test#>.
> _:list1 rdf:first :a; rdf:rest _:l2.
> _:l2 rdf:first :b; rdf:rest _:l3.
> _:l3 rdf:first :c; rdf:rest rdf:nil.
> _:list1 :p :o.
> (1 2 3) :p :o.


FILTER or QUERY

> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> @prefix list: <http://www.w3.org/2000/10/swap/list#>.
> @prefix : <http://example.org/test#>.
> {?LIST :p :o; rdf:first ?FIRST; list:last ?LAST} => {?LIST :first_element
> ?FIRST; :last_element ?LAST}.


With CWM I get

>              @prefix : <http://example.org/test#> .
>       ( 1
>         2
>         3 )
>          :first_element 1;
>          :last_element 3 .
>       ( :a
>         :b
>         :c )
>          :first_element :a;
>          :last_element :c .


which is the same as we get with EYE

> @prefix : <http://example.org/test#>.
> (:a :b :c) :first_element :a.
> (:a :b :c) :last_element :c.
> (1 2 3) :first_element 1 .
> (1 2 3) :last_element 3 .


Jos

-- https://josd.github.io/ <http://josd.github.io/>

**Comment by gkellogg** [comment]
> Well, just like list:last I implemented rdf:first also a built-in, but one that you can extend with your own triples Remember last Monday how we argued against such extensions like for instance (1 1) math:sum 3 but this is no problem for the built-ins rdf:first, rdf:rest, log:implies, etc, ... Well, to me it smells like a use versus mention bug, but one that we can deal with esp. because they are built-ins.

> Now could you try the following test case: GIVEN

```turtle
@Prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@Prefix list: <http://www.w3.org/2000/10/swap/list#>.
@Prefix : <http://example.org/test#>.
_:list1 rdf:first :a; rdf:rest _:l2.
_:l2 rdf:first :b; rdf:rest _:l3.
_:l3 rdf:first :c; rdf:rest rdf:nil.
_:list1 :p :o.
(1 2 3) :p :o.
```

> FILTER or QUERY

```turtle
@Prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@Prefix list: <http://www.w3.org/2000/10/swap/list#>.
@Prefix : <http://example.org/test#>.
{?LIST :p :o; rdf:first ?FIRST; list:last ?LAST} => {?LIST :first_element ?FIRST; :last_element ?LAST}.
```

With CWM I get

```turtle
@Prefix : <http://example.org/test#> .
( 1 2 3 ) :first_element 1; :last_element 3 .
( :a :b :c ) :first_element :a; :last_element :c .
```

which is the same as we get with EYE

```turtle
@Prefix : <http://example.org/test#>.
(:a :b :c) :first_element :a.
(:a :b :c) :last_element :c.
(1 2 3) :first_element 1 .
(1 2 3) :last_element 3 .
```

I get the following:

```turtle
@prefix : <http://example.org/test#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

(1 2 3) :p :o .
(:a :b :c) :first_element :a;
  :last_element :c;
  :p :o .
```

But, that is because I don't treat `rdf:first/rest` as builtins that can introspect a first-class list. It makes sense to do so, although IIRC, we haven't defined that particular behavior.

If I turn off the parser option that treats lists as first-class resources, I get the same results as you.

**Comment by gkellogg** [comment]
Interestingly, I was wrong about my interpretation as described above, as I had forgotten that I have a special interpretation for unbound existential variables.

Try the following:

```turtle
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix list: <http://www.w3.org/2000/10/swap/list#>.
@prefix : <http://example.org/test#>.

{ (1 2 3) list:last ?x } => { :lastMember :is ?x } .

{
  _:l3 rdf:first 3; rdf:rest rdf:nil .
  _:l2 rdf:first 2; rdf:rest _:l3 .
  _:l1 rdf:first 1; rdf:rest _:l2; list:last ?y
} => { :lastMember2 :is ?y } .
```

I get:

```turtle
:lastMember :is 3 .
:lastMember2 :is 3 .
```

Whereas before, I suggested that it would only have the first triple as a result.

Other than for pure semantics, my motivation for using first-class lists was to improve query performance (and understandability), as I was rapidly lost in the query patterns composed of all the rdf:first/rest combinations and found it easier to extend my pattern matching to encompass lists (much as I had previously done for embedded triples for RDF*).

**Comment by pchampin** [comment]
@gkellogg 
I understand the argument about performance and code readability.
But semantically, I see no problem. According to the [team submission](https://www.w3.org/TeamSubmission/n3/#lists):

> All lists exists

And the [N3 paper](https://arxiv.org/pdf/0711.1533.pdf) formalizes this through axioms:
```turtle
(?X).
{ ?L rdf:rest [ ] } => { [ ] rdf:first ?X; rdf:rest ?L }.
```
So the second form of your rule (with the b-node ladder) should match, regardless of the content of the knowledge base.

PS: granted, the axioms above are tricky. I tried to implement them with my own :first and :rest predicates (to avoid interaction with the builtin semantics of `rdf:first` and `rdf:rest`), and both EYE and CWM got stuck in infinite loops. I suspect they try to generate all possible lists, creating new nodes in the process, thus creating new lists...

**Comment by koo5** [comment]
@doerthe, apologies for not responding earlier. "AE rules" was my misspelling of "∀∃-rules", as outlined here:(https://www.sciencedirect.com/science/article/pii/S0004370211000397)

Before i try to explain my reasoner in detail, i think i will try to work on finishing one more, polished, version, hooking it up to the parser, and to the java test harness, and maybe getting some idea of what i'm talking about:) But i'm not sure if it matters that some implementation may be able to pass all list tests and give the exact same semantics for both forms of lists, if some won't.

I said: `such duality is incorrect`. I meant, incorrect under the proposition that an N3 list at any place can be replaced with a ladder and vice versa, and you should still get the exact same results. 

I understand why implementations do grant special "builtin powers" to lists that happened to be serialized as N3 lists, but the idea of controlling this through the syntax doesn't feel entirely right. Maybe we should have a distinction between the serialization/appearance, and between the "power" of the list? All proper lists without additional properties can be "compacted" into a "parsed N3 list form". Different implementations will find different sets of results to queries. So, shouldnt they try their best and just do this "compaction" automatically?
2) existing systems capable of processing quads will choke on this one weird new feature called "N3 list", which is a special new syntax whose only purpose is to notify reasoners that they should enable some builtin superpowers when encountering it. The problem will be that they will have to preserve the choice of the syntax, otherwise, a graph used as a query might yield different results before and after being stored in a triplestore. (?)

**Comment by william-vw** [comment]
@koo5 I'm a bit confused by some parts of your message:

> I said: `such duality is incorrect`. I meant, incorrect under the proposition that an N3 list at any place can be replaced with a ladder and vice versa, and you should still get the exact same results.

> I understand why implementations do grant special "builtin powers" to lists that happened to be serialized as N3 lists, but the idea of controlling this through the syntax doesn't feel entirely right. 

If I'm understanding correctly, you propose there **should not** be a difference in treatment between N3 lists and RDF first-rest ladders (possibly with [the caveat](https://github.com/w3c/N3/issues/12#issuecomment-669427773) discussed earlier)? (basing myself on some of your prior posts as well)

> Different implementations will find different sets of results to queries. 

They certainly shouldn't - at least, not as long as they follow the spec.

> 2) existing systems capable of processing quads will choke on this one weird new feature called "N3 list", which is a special new syntax whose only purpose is to notify reasoners that they should enable some builtin superpowers when encountering it. 

Well yes, I believe that is the goal of most builtins :-)

Note that a stop-gap (or even more permanent) solution here could be to simply add rules that convert RDF first-rest ladders to N3 lists (as per [Doerthe's post](https://github.com/w3c/N3/issues/12#issuecomment-478129474)).
